blockchain/state_manager/src/lib.rs

```diff
@@ -223,10 +365,229 @@ where
+                    }
+                    let error_msg = format!("found message with equal nonce as the one we are looking for (F:{:} n {:}, TS: `Error Converting message to Cid` n{:})", cid, message.sequence(), s.sequence());
+                    return Some(Err(Error::Other(error_msg)))
+                }
+                if s.sequence() < message.sequence() {
+                    return Some(Ok(None));
+                }
+
+                None
+            })
+            .next()
+            .unwrap_or_else(|| Ok(None))
+    }
+
+    /// WaitForMessage blocks until a message appears on chain. It looks backwards in the chain to see if this has already
+    /// happened. It guarantees that the message has been on chain for at least confidence epochs without being reverted
+    /// before returning.
+    pub async fn wait_for_message(
+        &self,
+        cid: &Cid,
+        confidence: i64,
+    ) -> Result<(Option<Arc<Tipset>>, Option<MessageReceipt>), Error> {
+        let mut subscribers = self.subscriber.clone().ok_or_else(|| {
+            Error::Other("State Manager not subscribed to tipset head changes".to_string())
+        })?;
+        let (sender, receiver) = oneshot::channel::<()>();
+        let message = chain::get_chain_message(self.get_block_store_ref(), cid)
+            .map_err(|err| Error::Other(format!("failed to load message {:}", err)))?;
+
+        let maybe_subscriber: Option<HeadChange> = subscribers.next().await;
+        let first_subscriber = maybe_subscriber.ok_or_else(|| {
+            Error::Other("SubHeadChanges first entry should have been one item".to_string())
+        })?;
+
+        let tipset = match first_subscriber {
+            HeadChange::Current(tipset) => tipset,
+            _ => {
+                return Err(Error::Other(format!(
+                    "expected current head on SHC stream (got {:?})",
+                    first_subscriber
+                )))
+            }
+        };
+        let maybe_message_reciept = self.tipset_executed_message(&tipset, cid, &message)?;
+        if let Some(r) = maybe_message_reciept {
+            return Ok((Some(tipset.clone()), Some(r)));
+        }
+
+        let mut candidate_tipset: Option<Arc<Tipset>> = None;
+        let mut candidate_receipt: Option<MessageReceipt> = None;
+        let mut back_tipset: Option<Tipset> = None;
+        let mut back_receipt: Option<MessageReceipt> = None;
+        async {
+            let (back_t, back_r) = self
+                .search_back_for_message(&tipset, &message)?
+                .ok_or_else(|| {
+                    Error::Other("State manager not subscribed to back search wait".to_string())
+                })?;
+            back_tipset = Some(back_t);
+            back_receipt = Some(back_r);
+            sender
+                .send(())
+                .map_err(|e| Error::Other(format!("Could not send to channel {:?}", e)))?;
+            Ok::<(), Error>(())
+        }
+        .await?;
```

Comment:
this will just compute this sequentially and await on it, would have to start a task to do this, and make sure it's cancelled if returned before it ends

Reply:
the issue was that the DB Associated Type is not a send it cannot be shared safely between threads. That was my initial struggle but it could benefit from the way futures are executed form the pool but I guess not. I don't really see the need for this unless search_back_message can be executed async since it is the one that is going to take a lot of execution time

Reply:
the only reason the db is a ref here is because it wasn't needed to remain and have cloned the `Arc`. If this does require concurrency (which is does if you want to match functionality) you should just change the parameter to an `Arc<DB>` and clone it for the async task that is doing this in parallel.

Calling await directly after the async block defeats the purpose of this